This fixes #394 and #395

jeroenbreen opened the pull request:
When an operator is deleted in the UI, this can have the effect that the corresponding parameter has no operators anymore or that the cps effect for certain elements has to be removed. 
When this happens we erase the parameter (394). If this is the last parameter of the rule, we erase the rule (395) (and update the administration of the rules with a higher index then the just deleted one).

graphicore commented:
@jeroenbreen the commit with this message "Merge remote-tracking branch 'metapolator/master'" shows that you got it a bit wrong. Instead of merging master you should do a "rebase". Looks like this, in your  `jeroenbreen:#395` branch for example, if `git@github.com:metapolator/metapolator.git` is the remote named upstream (`git remote add upstream git@github.com:metapolator/metapolator.git`) then:

1. Download the latest data: `git fetch upstream master`
2. Apply the latest changes to you branch, as if they happened before your work: `git rebase upstream/master`

jeroenbreen commented:
Ok, I did that: 
```
First, rewinding head to replay your work on top of it...
Applying: removed logs
Applying: added function to get all elements of a master
Applying: remove a parameter from a rule + remove rule + change rule indexes after removing a rule
Applying: uncommented remove parameter until parameter.erase is repaired
Applying: Remove a rule when its empty
```
So if I understand this correctly, that only changed the order of some things right, it didn't create diffs.
And I can keep this PR untouched?

graphicore commented:
In the end The commit that says "Merge remote-tracking branch 'metapolator/master' `b7e8ec2` from above should not be there.

The output of the rebase looks good, I think. If you check your logs, is `b7e8ec2` still in there? Did you push --force your branch after doing the rebase? Maybe we can remove that commit by hand???

davelab6 commented:
I don't think its a wise use of time to spend time fine tuning the commit
logs like this ;)

graphicore commented:
I think it's not hard to do and it is always wise to learn how your tools work. Also, we documented it. Where is there point at which an agreement between how things are done in a project are not wise anymore?

davelab6 commented:
I am partially joking around; you guys should run the engineering the way you feel is best and makes you happy, of course :)

jeroenbreen commented:
I think making me happy and doing things right with committing and stuff are 2 different roads ;)

jeroenbreen commented:
this is not good I guess: @graphicore 

```
commit 0d66cbee6763219d9706d7bf69923fc5fb4ee42c
Merge: b7e8ec2 0dc3d20
Author: Jeroen Breen <jeroen@innouveau.nl>
Date:   Fri Sep 4 15:20:41 2015 +0200

    Merge remote-tracking branch 'metapolator/master'

commit 0dc3d20436f38ea5564bb563162d45a0689cc527
Merge: 980e2ef 419fdd8
Author: Lasse Fister <commander@graphicore.de>
Date:   Thu Sep 3 23:54:40 2015 +0200

    Merge pull request #638 from graphicore/cps-updates_partial-dev-tool
    
    CPS engine updates and partial dev tool and some maintenance commits

commit 419fdd89795f73425d8d727106bf8b7cc018d6db
Author: Lasse Fister <commander@graphicore.de>
Date:   Thu Sep 3 21:59:11 2015 +0200

    Export metapolator apps for static deployments on a simple webserver
    
    Used to be called 'export-static-red-pill
:
commit 0d66cbee6763219d9706d7bf69923fc5fb4ee42c
Merge: b7e8ec2 0dc3d20
Author: Jeroen Breen <jeroen@innouveau.nl>
Date:   Fri Sep 4 15:20:41 2015 +0200

    Merge remote-tracking branch 'metapolator/master'
```

graphicore commented:
So these: "0dc3d20436f38ea5564bb563162d45a0689cc527 Merge pull request #638 from graphicore/cps-updates_partial-dev-tool" happen when we merge a pull request into master. We want to have them there.

These: "Merge remote-tracking branch 'metapolator/master'" We don't  want to have them, becuause they also go into master and that makes no sense there--why should I merge master into master. Using `rebase`, these messages never happen.

Here is a nice article: https://www.atlassian.com/git/tutorials/merging-vs-rebasing/

jeroenbreen commented:
so should I do that `force` thing? If so, can you quote please?

graphicore commented:
Yes. If you just did a `git rebase` on your feature branch `git push --force` is the way to update the remote on github, so that we can see it here in this PR. Before doing so, check if all of your work is still there (on your local version). It shouldn't disappear usually, but the `git push --force` changes your remote a lot and it may be harder to recover from any problems after.

Also, never use `git push --force` on a publicly shared repository, like metapolator/metapolator/master, because it's hard for others to sync then. We use the `rebase`/`push --force` combnation only on our own feature branches.

jeroenbreen commented:
ok think the 'Merge remote-tracking' are gone now.

graphicore commented:
yeah, looks like :smile:
